cs/src/comm/service/ServiceHost.cs

```diff
@@ -109,125 +113,108 @@ public void ValidateServiceMethods(IEnumerable<ServiceMethodInfo> serviceMethods
             logger.Site().Information("Deregistered {0}.", typeof(T).Name);
         }
 
-        public async Task<IMessage> DispatchRequest(
-            string methodName, ReceiveContext context, IMessage message, ConnectionMetrics connectionMetrics)
+        public async Task<IMessage> DispatchRequest(string methodName, ReceiveContext context, IMessage message)
         {
-            var totalTime = Stopwatch.StartNew();
             Stopwatch serviceTime = null;
-            var requestMetrics = StartRequestMetrics(methodName, connectionMetrics);
             logger.Site().Information("Got request [{0}] from {1}.", methodName, context.Connection);
 
-            ServiceMethodInfo methodInfo;
-
-            lock (dispatchTableLock)
+            try
             {
-                if (!dispatchTable.TryGetValue(methodName, out methodInfo))
+                ServiceMethodInfo methodInfo;
+                lock (dispatchTableLock)
+                {
+                    if (!dispatchTable.TryGetValue(methodName, out methodInfo))
+                    {
+                        var errorMessage = "Got request for unknown method [" + methodName + "].";
+
+                        logger.Site().Error(errorMessage);
+                        var error = new Error
+                        {
+                            message = errorMessage,
+                            error_code = (int) ErrorCode.MethodNotFound
+                        };
+                        return Message.FromError(error);
+                    }
+                }
+
+                if (methodInfo.CallbackType != ServiceCallbackType.RequestResponse)
                 {
-                    var errorMessage = "Got request for unknown method [" + methodName + "].";
+                    var errorMessage = "Method [" + methodName + "] invoked as if it were " +
+                                       ServiceCallbackType.RequestResponse + ", but it was registered as " +
+                                       methodInfo.CallbackType + ".";
 
                     logger.Site().Error(errorMessage);
                     var error = new Error
                     {
                         message = errorMessage,
-                        error_code = (int)ErrorCode.MethodNotFound
+                        error_code = (int) ErrorCode.InvalidInvocation
```

Comment:
This is Resharper's default, and we haven't overridden it in cs.sln.DotSettings, so I believe the space is what we want to stick to.